rectools/models/bert4rec.py

```diff
@@ -0,0 +1,578 @@
+        """Right truncation, left padding to session_max_len"""
+        x = np.zeros((len(batch), self.session_max_len))
+        for i, (ses, _) in enumerate(batch):
+            session = ses.copy()
+            session = session + [1]
+            x[i, -len(ses) :] = ses[-self.session_max_len :]
+        return torch.LongTensor(x)
+
+    def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
+        """TODO"""
+        sequence_dataset = SequenceDataset.from_interactions(dataset.interactions.df)
+        recommend_dataloader = DataLoader(
+            sequence_dataset,
+            batch_size=self.batch_size,
+            collate_fn=self._collate_fn_recommend,
+            num_workers=self.dataloader_num_workers,
+            shuffle=False,
+        )
+        return recommend_dataloader
+
+
+class PointWiseFeedForward(nn.Module):
+    """TODO"""
+
+    def __init__(self, n_factors: int, n_factors_ff: int, dropout_rate: float) -> None:
+        """TODO"""
+        super().__init__()
+        self.ff_linear1 = nn.Linear(n_factors, n_factors_ff)
+        self.ff_gelu = torch.nn.GELU()
+        self.ff_dropout = torch.nn.Dropout(dropout_rate)
+        self.ff_linear2 = nn.Linear(n_factors_ff, n_factors)
+
+    def forward(self, seqs: torch.Tensor) -> torch.Tensor:
+        """TODO"""
+        output = self.ff_gelu(self.ff_linear1(seqs))
+        fin = self.ff_linear2(self.ff_dropout(output))
+        return fin
+
+
+class BERT4RecTransformerLayers(TransformerLayersBase):
+    """TODO"""
+
+    def __init__(
+        self,
+        n_blocks: int,
+        n_factors: int,
+        n_heads: int,
+        dropout_rate: float,
+    ):
+        super().__init__()
+        self.n_blocks = n_blocks
+        self.multi_head_attn = nn.ModuleList(
+            [nn.MultiheadAttention(n_factors, n_heads, dropout_rate, batch_first=True) for _ in range(n_blocks)]
+        )
+        self.layer_norm1 = nn.ModuleList([nn.LayerNorm(n_factors) for _ in range(n_blocks)])
+        self.dropout1 = nn.ModuleList([nn.Dropout(dropout_rate) for _ in range(n_blocks)])
+        self.layer_norm2 = nn.ModuleList([nn.LayerNorm(n_factors) for _ in range(n_blocks)])
+        self.feed_forward = nn.ModuleList(
+            [PointWiseFeedForward(n_factors, n_factors * 4, dropout_rate) for _ in range(n_blocks)]
+        )
+        self.dropout2 = nn.ModuleList([nn.Dropout(dropout_rate) for _ in range(n_blocks)])
+        # self.dropout3 = nn.ModuleList([nn.Dropout(dropout_rate) for _ in range(n_blocks)])
+
+    def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: torch.Tensor) -> torch.Tensor:
+        """TODO"""
+        for i in range(self.n_blocks):
+            mha_input = self.layer_norm1[i](seqs)
+            # mha_output, _ =
+            # self.multi_head_attn[i](mha_input, mha_input, mha_input, attn_mask=attn_mask, need_weights=False)
```

Comment:
This is not handy at all :)
Here just pass the one you received in `forward` always.
And create a flag in the model if you want to create this mask or not. If not - pass it with Null value

Reply:
Also here for BERT we need to pass `~timeline_mask` as `key_padding_mask`

Reply:
fixed